### PR TITLE
FINERACT-1760: savings account externalId support

### PR DESCRIPTION
**Pull request by abraham-menyhart**
## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountRepository.java [hunk 1]
```diff
@@ -66,4 +66,7 @@ SavingsAccount findByIdAndDepositAccountType(@Param("accountId") Long accountId,
     Page<SavingsAccount> findByStatus(Integer status, Pageable pageable);
 
     SavingsAccount findByExternalId(String externalId);
+
+    @Query("SELECT sa.id FROM SavingsAccount sa WHERE sa.externalId = :externalId")
+    Long findIdByExternalId(@Param("externalId") String externalId);
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountRepository.java]: Please use `ExternalId` type here

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/SavingsAccountsTest.java [hunk 2]
```diff
@@ -80,18 +101,73 @@ void approveSavingsAccount() {
     }
 
     @Test
-    @Order(3)
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/SavingsAccountsTest.java]: Why did you remove the `activation` use case?

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please kindly see my comments!

**Comment by abraham-menyhart** [comment]
Thanks for your review @adamsaghy, I fixed the PR according to your comments. Can you please review it again?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountRepository.java [hunk 3]
```diff
@@ -65,5 +66,8 @@ SavingsAccount findByIdAndDepositAccountType(@Param("accountId") Long accountId,
 
     Page<SavingsAccount> findByStatus(Integer status, Pageable pageable);
 
-    SavingsAccount findByExternalId(String externalId);
+    SavingsAccount findByExternalId(ExternalId externalId);
+
+    @Query("SELECT sa.id FROM SavingsAccount sa WHERE sa.externalId = :externalId")
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountRepository.java:71]: This method would work without the `@Query` annotation as well! ;)

**Review by adamsaghy — APPROVED** [review]
LGTM

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountRepository.java (hunk 3 above)
Comment (abraham-menyhart) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountRepository.java:71]: Yes, nice catch, thanks!

**Comment by adamsaghy** [comment]
@abraham-menyhart  Run './gradlew :integration-tests:spotlessApply' to fix these violations.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/DepositAccountAssembler.java [hunk 4]
```diff
@@ -323,10 +324,10 @@ public SavingsAccount assembleFrom(final JsonCommand command, final AppUser subm
                     prodTermAndPreClosure);
 
             FixedDepositAccount fdAccount = FixedDepositAccount.createNewApplicationForSubmittal(client, group, product, fieldOfficer,
-                    accountNo, externalId, accountType, submittedOnDate, submittedBy, interestRate, interestCompoundingPeriodType,
-                    interestPostingPeriodType, interestCalculationType, interestCalculationDaysInYearType, minRequiredOpeningBalance,
-                    lockinPeriodFrequency, lockinPeriodFrequencyType, iswithdrawalFeeApplicableForTransfer, charges,
-                    accountTermAndPreClosure, accountChart, withHoldTax);
+                    accountNo, ExternalIdFactory.produce(externalId), accountType, submittedOnDate, submittedBy, interestRate,
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/DepositAccountAssembler.java]: Here we should generate one if nothing was provided: `ExternalIdFactory.create(...)`

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/DepositAccountAssembler.java [hunk 5]
```diff
@@ -341,7 +342,7 @@ public SavingsAccount assembleFrom(final JsonCommand command, final AppUser subm
                     prodRecurringDetail.recurringDetail());
 
             RecurringDepositAccount rdAccount = RecurringDepositAccount.createNewApplicationForSubmittal(client, group, product,
-                    fieldOfficer, accountNo, externalId, accountType, submittedOnDate, submittedBy, interestRate,
+                    fieldOfficer, accountNo, ExternalIdFactory.produce(externalId), accountType, submittedOnDate, submittedBy, interestRate,
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/DepositAccountAssembler.java]: Here we should generate one if nothing was provided: `ExternalIdFactory.create(...)`

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccount.java [hunk 6]
```diff
@@ -1662,10 +1664,10 @@ public void modifyApplication(final JsonCommand command, final Map<String, Objec
             this.accountNumber = StringUtils.defaultIfEmpty(newValue, null);
         }
 
-        if (command.isChangeInStringParameterNamed(SavingsApiConstants.externalIdParamName, this.externalId)) {
+        if (command.isChangeInStringParameterNamed(SavingsApiConstants.externalIdParamName, this.externalId.getValue())) {
             final String newValue = command.stringValueOfParameterNamed(SavingsApiConstants.externalIdParamName);
             actualChanges.put(SavingsApiConstants.externalIdParamName, newValue);
-            this.externalId = StringUtils.defaultIfEmpty(newValue, null);
+            this.externalId = ExternalIdFactory.produce(newValue);
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccount.java:1671]: Here we should generate one if nothing was provided: `ExternalIdFactory.create(...)`

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountAssembler.java [hunk 7]
```diff
@@ -323,10 +324,10 @@ public SavingsAccount assembleFrom(final JsonCommand command, final AppUser subm
         }
 
         final SavingsAccount account = SavingsAccount.createNewApplicationForSubmittal(client, group, product, fieldOfficer, accountNo,
-                externalId, accountType, submittedOnDate, submittedBy, interestRate, interestCompoundingPeriodType,
-                interestPostingPeriodType, interestCalculationType, interestCalculationDaysInYearType, minRequiredOpeningBalance,
-                lockinPeriodFrequency, lockinPeriodFrequencyType, iswithdrawalFeeApplicableForTransfer, charges, allowOverdraft,
-                overdraftLimit, enforceMinRequiredBalance, minRequiredBalance, maxAllowedLienLimit, lienAllowed,
+                ExternalIdFactory.produce(externalId), accountType, submittedOnDate, submittedBy, interestRate,
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountAssembler.java]: Here we should generate one if nothing was provided: `ExternalIdFactory.create(...)`

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please see my comments!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccount.java (hunk 6 above)
Comment (abraham-menyhart) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccount.java:1671]: Since it is an Entity class, and the create method is not static, I should pass the `ExternalIdFactory` dependency as a method parameter, etc. So it won't be very pretty I think. Is it a business requirement? In the other cases, I'll implement your change request, only this case I'm a bit hesitant.

Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccount.java:1671]: Till this logic is not extracted out from the entity you might wanna use this:
```
            ExternalId externalId = ExternalIdFactory.produce(newValue);
            if (externalId.isEmpty() && TemporaryConfigurationServiceContainer.isExternalIdAutoGenerationEnabled()) {
                externalId = ExternalId.generate();
            }
```

Comment (abraham-menyhart) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccount.java:1671]: Thanks! Updated the PR according to this.
